### ISSUE TITLE
Translucent blocks do not render correctly (fixed in Sodium 0.6)

### ISSUE THREAD

**Issue by RdSt14** [jellysquid3/Sodium#38]
### Expected Behavior
Water will be visible inside glass stained pipes, commonly used in water elevators

### Actual Behavior
Water is not visible through stained glasses, although items can be seen flowing

### Reproduction Steps
Make a water tube, doesn't matter if its flowing or a water elevator type.

### Attachments
https://cdn.discordapp.com/attachments/389371678859460611/726798724206166066/received_707934516451025.png

### System Information
- Java Version: [Java 8 Update 251]
- CPU: [Intel Core i5-7200U]
- GPU: [Intel HD Graphics 620]

**Comment by Xephobia** [comment]
also it show other weird thing like the sky rendering black instead of white or stained glass can't show other stained glasses behind them

**Comment by jellysquid3** [comment]
Sodium does not implement quad sorting for translucent block faces, leading to all kinds of translucency issues. I'll be making this a tracking issue for other problems caused by it.

**Comment by muzikbike** [comment]
Can confirm all kinds of weird behaviour with translucent blocks.
![2020-07-15_20 38 01](https://user-images.githubusercontent.com/52297970/87610303-ea1c4180-c6fc-11ea-8528-6985309fd0f7.png)
![2020-07-15_20 40 25](https://user-images.githubusercontent.com/52297970/87610310-ebe60500-c6fc-11ea-8e83-0d6332963757.png)


You can even see chunk borders in ice:
![2020-07-15_20 44 41](https://user-images.githubusercontent.com/52297970/87610318-f0aab900-c6fc-11ea-94f1-7df24ee95d9a.png)
![2020-07-15_20 44 46](https://user-images.githubusercontent.com/52297970/87610320-f1dbe600-c6fc-11ea-8905-8acde6031e3a.png)

**Comment by catgirldoll** [comment]
## More images of the issue

no rendering issue when water rendering isn't effecting the block
![2020-07-17_18 25 40](https://user-images.githubusercontent.com/20545957/87841575-1c43b580-c85b-11ea-9b38-3d276bdb3062.png)

Only top texture on the block renders through water
![2020-07-17_18 08 09](https://user-images.githubusercontent.com/20545957/87841567-0e8e3000-c85b-11ea-941e-897ffd7c6bcf.png)

As a result of the issue above, you cannot see certain transparent blocks above you while inside of water
![2020-07-17_18 12 42](https://user-images.githubusercontent.com/20545957/87841570-14841100-c85b-11ea-86a5-49edaa057d03.png)

Issue applies to flowing water as well
![2020-07-17_18 14 41](https://user-images.githubusercontent.com/20545957/87841572-15b53e00-c85b-11ea-98db-0480b1754bb6.png)

**Comment by KrisCris** [comment]
Also, portals are invisible behind glass.

**Comment by James103** [comment]
See [this video by Xisumavoid](https://youtu.be/Jqp0r2SBdBM?t=80) for another example of this issue with Honey Blocks (they are almost as see-through as glass, even though they shouldn't be that translucent).

With Sodium:
![image](https://user-images.githubusercontent.com/37945304/88679340-f61ce180-d0b4-11ea-877d-011ceaed0a90.png)

Without Sodium:
![image](https://user-images.githubusercontent.com/37945304/88679398-03d26700-d0b5-11ea-81df-f663dc2ba715.png)

**Comment by zLauch** [comment]
Nether Portals also not showing up 
![2020-08-28_20 09 29](https://user-images.githubusercontent.com/44842475/91601860-a73fc180-e96a-11ea-8d4a-669c443c2c44.png)
![2020-08-28_20 10 56](https://user-images.githubusercontent.com/44842475/91601876-ac9d0c00-e96a-11ea-996e-9152b681bf85.png)

**Comment by lordofpipes** [comment]
Just in case this is useful, not sure if I'm rehashing what's already known -

I'm getting same rendering as @zLauch on f577eb2be13b93d8e584aaa7cc329680e0deaaf3 (and also on ishland's 1.16.2 support PR, though evidently that's not the actual cause) but not 0.1.0. Hardware is AMD Mullins, happens with all of sodium's rendering modes.

```
Java is version 1.8.0_262, using 64-bit architecture.

AMD A8-7410 APU with AMD Radeon R5 Graphics
Advanced Micro Devices, Inc. [AMD/ATI] Mullins [Radeon R4/R5 Graphics] (rev 45)

Subsystem: Dell Device 06bf

Kernel driver in use: amdgpu

OpenGL version string: 4.6 (Compatibility Profile) Mesa 20.1.6
```

**Comment by Jaekr** [comment]
Translucent blocks are even more broken,
Here is what I tested.

Information of the system used for the test:
```
Java version: 14.0.2 64 bits (tested with java 8 as well)
GPU: AMD RX 590
Driver: AMDGPU
OpenGL version: Mesa 20.1.7
Distro: ArchLinux
```
Version of Sodium: #250 latest commit as 09/06/2020
Note: you can click on thumbnails to open the full-sized original screenshot.

---
# Reference view (without Sodium):
[![Default](https://img.tedomum.net/data/thumb/2020-09-06_12.53.51-1478a9.png)](https://img.tedomum.net/data/2020-09-06_12.53.51-1478a9.png)

---
# Oneshot (GL 2.0)
[![Translucent blocks are broken](https://img.tedomum.net/data/thumb/2020-09-06_12.47.55-912f16.png)](https://img.tedomum.net/data/2020-09-06_12.47.55-912f16.png)

---
# Multidraw (GL 4.3)
[![Even more broken](https://img.tedomum.net/data/thumb/2020-09-06_12.55.43-f95b99.png)](https://img.tedomum.net/data/2020-09-06_12.55.43-f95b99.png)

---
# Oneshot (GL 3.3)
[![Same](https://img.tedomum.net/data/thumb/2020-09-06_12.57.13-e9b172.png)](https://img.tedomum.net/data/2020-09-06_12.57.13-e9b172.png)

**Comment by Hnaguski** [comment]
Not sure if this is helpful/already known but it seems like on chunk borders translucency effects are less broken? I've only tested it with honey/slime blocks but it seems like when you have two blocks on either side of a chunk border the middle part gets rendered correctly. The image should be a bit better at explaining.
![2020-09-27_20 20 38](https://user-images.githubusercontent.com/29007712/94384115-09711b00-00ff-11eb-9bfc-515887af9c62.png)

**Comment by KraXen72** [comment]
when looking throught stained glass, air and then more stained glass, parts of it are missing:
![2021-03-03_12 52 49](https://user-images.githubusercontent.com/21956756/109801982-7db86900-7c1f-11eb-920e-5576e6aa69c1.png)

**Comment by Boobies** [comment]
> Not sure if this is helpful/already known but it seems like on chunk borders translucency effects are less broken? I've only tested it with honey/slime blocks but it seems like when you have two blocks on either side of a chunk border the middle part gets rendered correctly. The image should be a bit better at explaining.
> ![2020-09-27_20 20 38](https://user-images.githubusercontent.com/29007712/94384115-09711b00-00ff-11eb-9bfc-515887af9c62.png)

Note that this issue affects subchunk borders, not just chunk borders. You'll see the same effect on the vertical axis, if you cross borders.

**Comment by edwloef** [comment]
![Screenshot from 2021-06-27 16-08-22](https://user-images.githubusercontent.com/75493688/123547662-ea70e980-d761-11eb-8daa-69dc70c179f3.png)

The issues with slime and honey blocks seem to be at least partly fixed in the 1.17 beta version.

**Comment by Youmiel** [comment]
1.17 beta version "seems" fix this issue, but in some cases it's still werid(For example the slime block behind white stained glass).
![ZV05LX_`HLW$34}Y8@O) F8](https://user-images.githubusercontent.com/70473080/123548262-a3511b80-d796-11eb-8a46-815542e5a118.png)

**Comment by JumperREDACTED** [comment]
![2021-09-26_21 45 31](https://user-images.githubusercontent.com/83337894/134822593-424e1d24-7c02-4cdc-8299-3db0e80fc36e.png)
![2021-09-26_21 45 52](https://user-images.githubusercontent.com/83337894/134822598-be37266e-b583-4bc1-89d2-05a8d42aaf28.png)
more pics ig, havent seen anyone mention how borked kelp is (except for the topmost part for some reason?)

**Comment by Kyrinon** [comment]
I notice this bug while viewing a column of water threw a multichunk plane of glass and notice it only effected the water and stained glass in the same sub chunk
![2021-10-13_22 39 53](https://user-images.githubusercontent.com/5389122/137242461-de679ba4-1780-477a-a8a1-ba176f31dd8a.png)
.

**Comment by LJNeon** [comment]
I noticed the kelp and water appear to render incorrectly with many blocks, not just glass. This was on version 0.3.2+build.7
![image](https://user-images.githubusercontent.com/23249107/138038141-0e0c2bed-4565-4c8a-b761-bdfa1d641cf5.png)
![image](https://user-images.githubusercontent.com/23249107/138038192-22cfd2dc-cfcc-4c0e-bcd3-329145e0f47f.png)
![image](https://user-images.githubusercontent.com/23249107/138038328-1af4c381-2865-4300-84f2-143c64deb494.png)

**Comment by LJNeon** [comment]
Also, all one-block tall ice appears to not render any submerged parts.
![image](https://user-images.githubusercontent.com/23249107/138038671-9ada4a4f-81a5-49c8-ab45-2994d6d71d6c.png)

**Comment by silversquirl** [comment]
This bug is:
- issues related to interactions between multiple layers semi-transparent blocks, such as water, ice, stained glass, nether portals, etc.

This bug is not:
- issues related to blocks with clip transparency, such as glass
- issues related to one layer of translucency, such as the water + kelp rendering issues reported in #932

# We know a lot about this bug already. Please refrain from posting unless you have something new to add.

**Comment by TimGoll** [comment]
I want to add to this issue two further things that I noticed:

1. only happens with stained glass
2. weird things happen on chunk borders

Here are two pictures from 1.18.1 (0.4.0-alpha5+build.9)

![image](https://user-images.githubusercontent.com/13639408/149153306-eaa15f5e-eb2c-449f-af4f-bbc3c9990623.png)
![image](https://user-images.githubusercontent.com/13639408/149153319-c0c29995-caa1-496b-94ce-c0dee04d677b.png)

**Comment by AugustusZebulon** [comment]
![image](https://user-images.githubusercontent.com/7438458/150612204-390e7bc5-347a-490f-83c7-ba7de15055ad.png)

![image](https://user-images.githubusercontent.com/7438458/150612317-8c4f3a13-3fb7-4e64-bb51-5ce5d4ac7513.png)
![image](https://user-images.githubusercontent.com/7438458/150612375-d443f7b3-c4bf-439b-8955-f2713a89b413.png)


Water in a water column renders with strips and breaks off directly on the borders it seems like. 
This changes while you move through the column where you can see random vertical and horizontal "panes" of darker water.

**Comment by goobyq** [comment]
> ![image](https://user-images.githubusercontent.com/7438458/150612204-390e7bc5-347a-490f-83c7-ba7de15055ad.png)
> 
> ![image](https://user-images.githubusercontent.com/7438458/150612317-8c4f3a13-3fb7-4e64-bb51-5ce5d4ac7513.png) ![image](https://user-images.githubusercontent.com/7438458/150612375-d443f7b3-c4bf-439b-8955-f2713a89b413.png)
> 
> Water in a water column renders with strips and breaks off directly on the borders it seems like. This changes while you move through the column where you can see random vertical and horizontal "panes" of darker water.
![java_oz3Hll8vpO](https://user-images.githubusercontent.com/24848664/159813708-cc790e86-3bfd-4d62-bf97-903f31dc9f38.png)
Not sure if it's been referenced elsewhere but on subchunks it adds slices/faces to water

**Comment by Yadokingu** [comment]
I have the same problem with glass not rendering through water.
![2022-04-16_11 53 30](https://user-images.githubusercontent.com/11133215/163670618-014867bf-cb94-48f9-83b9-03ae9fab3563.png)
![2022-04-16_11 54 47](https://user-images.githubusercontent.com/11133215/163670619-f3d78a60-33b0-43b7-bae1-6617eeb71eb9.png)

**Comment by RedstoneManiac3Xbox** [comment]
is there an actual implementable fix for this yet? or am I gonna be stuck not viewing the waterfalls in my terrarium?

**Comment by catgirldoll** [comment]
> is there an actual implementable fix for this yet? or am I gonna be stuck not viewing the waterfalls in my terrarium?

No, it's a WIP. See: [pull request](https://github.com/CaffeineMC/sodium-fabric/pull/963)

**Comment by nonetrix** [comment]
same issue on NVIDIA on Linux

**Comment by FreezePhoenix** [comment]
Don't know if anyone has noticed - but this only affects things that are placed in Translucent render layers. This includes water, tinted glass, stained glass, honey, slime, etc. It would appear that faces are simply sorted by some insertion order.
![2022-07-19_14 28 24](https://user-images.githubusercontent.com/24441367/179822957-29c551d3-7816-438a-8fa7-b1d4bb7c145d.png)
![2022-07-19_14 28 07](https://user-images.githubusercontent.com/24441367/179822967-6556681a-4be6-4979-8eea-e5fa10f501bc.png)
![2022-07-19_14 28 17](https://user-images.githubusercontent.com/24441367/179822975-a28dcf72-f9f5-48cf-b04f-fde30d43ef56.png)

**Comment by JaehaerysWise** [comment]
Posted another issue about water rendering issue in stained glass, but I feel like I should add it here as well
First pic (with continuity). Second pic (without continuity)
![2022-07-20_19 11 51](https://user-images.githubusercontent.com/109792039/180402555-feee010a-d6fa-4bca-8b80-0ea9d133e966.png)
![2022-07-20_19 31 40](https://user-images.githubusercontent.com/109792039/180402746-b366dcac-123d-4f7f-8745-dd58c0de6e31.png)

**Comment by silversquirl** [comment]
# Please stop!

...adding yet more images of the same thing

This bug is being actively worked on, and we **don't need any more information** about it.
Instead, sit down, relax, and wait patiently.

Thanks!

**Comment by jellysquid3** [comment]
This problem will be resolved in Sodium 0.6 with the introduction of #2016.
